pages/form/index.tsx

```diff
@@ -0,0 +1,192 @@
+import React, { FormEvent, useState } from "react";
+import axios from "axios";
+import InputForm from "../../components/registration-form/InputForm";
+
+export type FieldName =
+  | "email"
+  | "firstName"
+  | "lastName"
+  | "password"
+  | "passwordConfirm";
+
+export type FieldState = {
+  value: string;
+  error: string | null;
+};
+
+export type FormState = {
+  [fieldName: string]: FieldState;
+};
+
+const initialFieldState: FieldState = {
+  value: "",
+  error: null,
+};
+
+const initialFormState: FormState = {
+  email: initialFieldState,
+  firstName: initialFieldState,
+  lastName: initialFieldState,
+  password: initialFieldState,
+  passwordConfirm: initialFieldState,
+};
+
+const checkEmail = (email: string) => {
+  return /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email);
+};
+
+const updateError = (
+  previousState: FormState,
+  fieldName: FieldName,
+  error: string | null
+) => {
+  return {
+    ...previousState,
+    [fieldName]: { ...previousState[fieldName], error },
+  };
+};
+
+const RegisterForm = () => {
+  const [state, setState] = useState(initialFormState);
+  const { email, firstName, lastName, password, passwordConfirm } = state;
+
+  const handleSubmit = async (event: FormEvent) => {
+    event.preventDefault();
+
+    let currentState = { ...state };
+
+    currentState = updateError(
```

Comment:
i would probably do a bit more here...when you see yourself doing something exactly the same 3 times, it's likely an opportunity to refactor (then you need to decide whether it's worth refactoring or not)

also use `email.value?.trim()` to trim extra spaces around the input

```
function validateEmail(email: string) {...}
function validateFirstName(firstName: string) {...}
function validateLastName(lastName: string) {...}
function validatePassword(password: string) {...}
function validatePasswordConfirm(passwordConfirm: string, formState: FormState) {...}

const fieldValidations = {
  email: validateEmail,
  firstName: validateFirstName,
  ...
};

const submitForm = async (validForm: FormState) => {
  // submit thru axios
  setState(initialState);
};

const handleSubmit = async (...) => {
  const errors = Object.entries(fieldValidations)
    .map(([key, validationFn]) => [key, validationFn(state[key], state)])
    ;

    const hasError =  errors.some(([key, reason]) => !!reason);
    
    // update all field states, note that reason can be undefined
    const newState = errors.reduce(([key, reason], currentState) => updateError(currentState, key, reason), state);

    if (hasError) {
      setState(newState);
    } else {
      await submitForm();
    }
}
```

This way you can test the validation functions separately from the logic that handles submit button is clicked.